src/lib/demux.rs

```diff
@@ -421,6 +421,20 @@ where
             let mut output_reads = vec![];
             let mut qual_counter = BaseQualCounter::new();
 
+            // Ensure that all the reads have the same read name
+            ensure!(!zipped_reads.is_empty(), "Bug: no records found");
+            let first_read_name =
+                zipped_reads[0].head().splitn(2, |c| *c == b' ').next().unwrap_or(b"");
+            for read in &zipped_reads {
+                let cur_read_name = read.head().splitn(2, |c| *c == b' ').next().unwrap_or(b"");
+                ensure!(
+                    first_read_name == cur_read_name,
+                    "Read names did not match: {:?} != {:?}",
+                    String::from_utf8_lossy(first_read_name),
+                    String::from_utf8_lossy(cur_read_name)
+                );
+            }
```

Comment:
I think you could do this more efficiently as follows:

```rust
            let first_head = zipped_reads[0].head();
            let end_index = zipped_reads[0].head().find_byte(b' ').unwrap_or(first_head.len());
            for read in zipped_reads.iter().dropping(1) {
                let head = read.head();
                let ok = head.len() >= end_index && first_head[0..end_index] == head[0..end_index];
                ensure!(ok, ...);
            }
```

This calculates the region of the first header that is the read name, and then for subsequent reads checks they have at least that many characters, and that they match.  This avoids scanning all but the first read, and I think it does that more efficiently too.  For bonus points you could also double-check that each read header stops after `end_index` or that `head[end_index] == b' '`.

Reply:
For my future self:

```rust
let head = record.head();
let ok = head.len() == end_index || (head.len() > end_index && head[end_index] == b' ');
let ok = ok && first_head[0..end_index] == head[0..end_index];
```